sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DebugAssertHasNoSideEffects.cs

```diff
@@ -0,0 +1,92 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DebugAssertHasNoSideEffects : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3346";
+        private const string MessageFormat = "Expressions used in 'Debug.Assert' should not produce side effects.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<string> sideEffectWords = new HashSet<string>
+        {
+            "remove", "delete", "put", "set", "add", "pop", "update", "retain",
+            "insert", "push", "append", "clear", "dequeue", "enqueue", "dispose"
+        };
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var invokedMethodSyntax = c.Node as InvocationExpressionSyntax;
+
+                    if (GetMethodSymbol(c, invokedMethodSyntax).IsDebugAssert() &&
+                        invokedMethodSyntax
+                            .DescendantNodes()
+                            .OfType<InvocationExpressionSyntax>()
+                            .Select(GetIdentifierName)
+                            .Any(text => text != null &&
+                                    sideEffectWords.Overlaps(text.SplitCamelCaseToWords())))
+                    {
+                        c.ReportDiagnostic(Diagnostic.Create(rule, invokedMethodSyntax.ArgumentList.GetLocation()));
+                    }
+                },
+                SyntaxKind.InvocationExpression);
+        }
+
+        private static string GetIdentifierName(InvocationExpressionSyntax invocation)
+        {
+            var memberAccess = invocation.Expression as MemberAccessExpressionSyntax;
+
+            if (memberAccess != null)
+            {
+                return memberAccess.Name?.Identifier.ValueText;
+            }
+
+            var memberBinding = invocation.Expression as MemberBindingExpressionSyntax;
+            return memberBinding.Name?.Identifier.ValueText;
```

Comment:
If it can then you need to handle it. If it can't you might want to use direct cast to make it clear.